src/java.base/share/classes/java/math/BigDecimal.java

```
@@ -2169,198 +2158,110 @@ public BigDecimal sqrt(MathContext mc) {
+                BigInteger workingInt = working.toBigInteger();
+
+                BigInteger sqrt;
+                long resultScale = normScale >> 1;
+                if (mc.roundingMode == RoundingMode.DOWN || mc.roundingMode == RoundingMode.FLOOR) { // No need to round
+                    sqrt = workingInt.sqrt();
+                } else { // Round sqrt with the specified settings
+                    BigInteger[] sqrtRem = workingInt.sqrtAndRemainder();
+                    sqrt = sqrtRem[0];
+
+                    boolean increment = false;
+                    if (halfWay) { // half-way rounding
+                        // remove the one-tenth digit
+                        BigInteger[] quotRem10 = sqrt.divideAndRemainder(BigInteger.TEN);
+                        sqrt = quotRem10[0];
+                        resultScale--;
+
+                        int digit = quotRem10[1].intValue();
+                        if (digit > 5) {
+                            increment = true;
+                        } else if (digit == 5) {
+                            if (mc.roundingMode == RoundingMode.HALF_UP
+                                    || mc.roundingMode == RoundingMode.HALF_EVEN && sqrt.testBit(0)
+                                    // Check if remainder is non-zero
+                                    || sqrtRem[1].signum != 0 || working.compareTo(new BigDecimal(workingInt)) != 0) {
+                                increment = true;
+                            }
                         }
-                        result = result.subtract(ulp);
-                    }
-                    break;
-
-                case UP:
-                case CEILING:
-                    // Check if too small
-                    if (result.square().compareTo(this) < 0) {
-                        result = result.add(result.ulp());
+                    } else { // mc.roundingMode == RoundingMode.UP || mc.roundingMode == RoundingMode.CEILING
```

It would be more robust to make all enum constant explicit, and reserve the `else` to throw a `RuntimeException` of some sort, e.g., `MatchException`. Should an additional enum constant be added to `RoundingMode` in the future, this code will then throw instead of silently producing a wrong result.

If possible at all, it would be preferable to use a `switch` rather than `if`s.

---

While you are at it, it would be useful to add one test for each of the paths in the new algorithm (including the exception cases) in the existing [SquareRootTests](https://github.com/openjdk/jdk/blob/master/test/jdk/java/math/BigDecimal/SquareRootTests.java) This helps to identify future regressions when the code is later modified for any reason.